### PR TITLE
Create annotation ExperimentalApi and Internal.

### LINKED ISSUE #737: Remove ChannelImpl.ping (or possible mark it experimental)
It isn't clear that ping will be part of the long term gRPC interface (and is easy to add back in later if needed).  Either remove it or mark it experimental.

### PR DESCRIPTION
**Pull request by zhangkun83**
Resolves #737

### CODE REVIEW

**Comment by ejona86** [comment]
@zhangkun83 LGTM. I was thinking we would have an annotation of some sort. This is still better than nothing, but we may want to add an annotation later.

**Comment by zhangkun83** [comment]
Added `@ExperimentalApi` to:                                                            
- Context (resolves #740)
- ChannelImpl.ping (resolves #737)
- AbstractChannelBuilder.ChannelEssentials (#733)

@ejona86 PTAL

**Comment by carl-mastrangelo** [comment]
@zhangkun83 alternate option:  Would the @Beta tag from guava be appropriate instead?

**Comment by zhangkun83** [comment]
> @zhangkun83 alternate option: Would the @ Beta tag from guava be appropriate instead?

@ejona86 I have an impression that we want to get rid of Guava dependency. Is it the case?

**Comment by ejona86** [comment]
> @ejona86 I have an impression that we want to get rid of Guava dependency. Is it the case?

Unclear. We are riding the fence on that one, and have mainly just been avoiding adding dependencies in the API. If we swapped from Guava's Beta to our own annotation in the future, it is hard to say whether that is an API breakage. If it's not an API breakage, then using Guava's Beta would be fine because we still have the freedom to not depend on Guava should we want to.

File: core/src/main/java/io/grpc/ExperimentalApi.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc;
+
+/**
+ * Indicates a public API that can change at any time, and has no guarantee of API stability and
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ExperimentalApi.java]: Can it be said that this is a form of documentation, and should not be used by other projects? i.e., this is not part of our API.

File: core/src/main/java/io/grpc/ExperimentalApi.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * <p>How to use it:
+ * <ol>
+ * <li>This annotation is used only on public API. Internal interfaces should not use it.</li>
+ * <li>After gRPC has gained API stability, this annotation can only be added to new API. Adding it
+ * to an existing API is considered API-breaking.</li>
+ * <li>Removing this annotation from an API gives it stable status.</li>
+ * </ol>
+ */
+public @interface ExperimentalApi {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ExperimentalApi.java]: @ Documented, so it shows up on JavaDoc?

Comment (ejona86) [inline core/src/main/java/io/grpc/ExperimentalApi.java]: Do we want CLASS @ Retention, or is SOURCE appropriate?

File: core/src/main/java/io/grpc/ExperimentalApi.java (hunk 1 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/ExperimentalApi.java]: How about annotating it with `@Internal`? `@Internal` itself will be annotated with `@Internal` too.

Comment (ejona86) [inline core/src/main/java/io/grpc/ExperimentalApi.java]: That sounds fine.

File: core/src/main/java/io/grpc/ExperimentalApi.java (hunk 2 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/ExperimentalApi.java]: I think SOURCE is appropriate, as we are marking them both as `@Internal`, thus we don't expect any tool to check for them in class files.

**Comment by zhangkun83** [comment]
- Add `@Internal` annotation                                                  
- Both `@ExperimentalApi` and `@Internal` are annotated `@Internal`      
- Annotate `@Internal` to `package io.grpc.internal`                               
- AbstractChannelBuilder.ChannelEssentials is annotated `@Internal`                        
- Add `package-info.java` to `io.grpc.inprocess` and `io.grpc.internal`.

@ejona86 PTAL.

**Comment by ejona86** [comment]
@zhangkun83 LGTM

**Pull request merged**